torch_geometric/loader/link_neighbor_loader.py

```diff
@@ -0,0 +1,146 @@
+
+        batch_size = len(index)
+
+        # take start and end node from each edge then deduplicate
+        node_index = torch.cat([row, col], dim=0)
+        node_index = torch.unique(node_index)
+
+        # get sampled graph
+        node, row, col, edge, _ = super().__call__(node_index)
+        return node, row, col, edge, batch_size, index
+
+
+class LinkNeighborLoader(DataLoader):
+    def __init__(
+        self,
+        data: Data,
+        num_neighbors: NumNeighbors,
+        input_edges: InputEdges = None,
+        input_edge_labels: OptTensor = None,
+        replace: bool = False,
+        directed: bool = True,
+        transform: Callable = None,
+        force_order: bool = False,
+        **kwargs,
+    ):
+
+        # Remove for PyTorch Lightning:
+        if 'dataset' in kwargs:
+            del kwargs['dataset']
+        if 'collate_fn' in kwargs:
+            del kwargs['collate_fn']
+
+        self.data = data
+
+        # Save for PyTorch Lightning < 1.6:
+        self.num_neighbors = num_neighbors
+        self.input_edges_idx = self._get_input_edges_idx(input_edges)
+        self.input_edge_labels = self._get_edge_labels(input_edge_labels)
+        self.replace = replace
+        self.directed = directed
+        self.transform = transform
+        self.neighbor_sampler = LinkNeighborSampler(
+            data, num_neighbors, replace, directed, force_order,
+            share_memory=kwargs.get('num_workers', 0) > 0)
+
+        super().__init__(self.input_edges_idx,
+                         collate_fn=self.neighbor_sampler, **kwargs)
+
+    def _get_iterator(self) -> Iterator:
+        return DataLoaderIterator(super()._get_iterator(), self.transform_fn)
+
+    def transform_fn(self, out: Any) -> Tuple[Data, torch.Tensor]:
+        node, row, col, edge, batch_size, index = out
+        data = filter_data(self.data, node, row, col, edge,
+                           self.neighbor_sampler.perm)
+
+        data.batch_size = batch_size
+        data = data if self.transform is None else self.transform(data)
+
+        labels = self.input_edge_labels[index]
+        edges = self.data.edge_index.T[torch.Tensor(self.input_edges_idx).type(
+            torch.long)[index]]
+
+        data.edge_label_index = edges
+        data.edge_labels = labels
+        return data
+
+    def _get_input_edges_idx(self, input_edges: InputEdges):
+        if isinstance(self.data, Data):
+            if input_edges is None:
+                return range(self.data.num_edges)
+
+            if isinstance(input_edges, EdgeType):
+                raise RuntimeError("`input_edges` cannot be string for non"
+                                   "hetrogenous graphs")
+
+            input_size = input_edges.size()
+
+            if len(input_size) == 1 and input_edges.dtype == torch.bool:
+                return mask_to_index(self.input_edge_idx)
+
+            if len(input_edges.size()) == 2 and input_edges.size()[0] == 2:
```

Comment:
Restart the conversation on the input specification here: The reason I initially changed the input name to `input_edge_idx` was so that it was clear we wouldn't do this (an inverse mapping from edges to their index). We need the index eventually because we need to recover the edge attributes in the transform function.

I'd be a fan of removing this personally and assuming its an index that is provided to this class - Although it would be nice to allow people to provide edges that don't exist and sample the graph for them, its also confusing what to do for the edge attributes in that case.

WDYT @rusty1s? Open to other approaches.

Reply:
I am not really sure I fully understand. Why do we need to recover edge attributes of supervision edges (that potentially do not even exist)?

In case you require an index to this loader, this means that you can only sample links that actually exist (i.e. which are "indexable" by `edge_index[idx]`), which does not seem to be the preferred solution. Ultimately, we want to infer missing links here as well.

Reply:
Got your point. I was being stupid.

I've updated the PR (it's still messy, but hopefully we're more aligned on the design). The only part I'm uncertain about is on line 85. Will open a conversation there.